python/taichi/lang/ops.py

```diff
@@ -373,22 +373,34 @@ def round(x):  # pylint: disable=redefined-builtin
 
 
 @unary
-def floor(x):
+def _floor(x):
+    return _unary_operation(_ti_core.expr_floor, math.floor, x)
+
+
+def floor(x, dtype=float):
     """Return the floor of the input, element-wise.
     The floor of the scalar `x` is the largest integer `k`, such that `k <= x`.
+
     Args:
         x (Union[:mod:`~taichi.types.primitive_types`, :class:`~taichi.Matrix`]): \
             Input scalar or matrix.
+
+        dtype: (:mod:`~taichi.types.primitive_types`): the returned type, default to `float` (current `default_fp`).
+
     Returns:
-        The floor of each element in `x`, with float type.
+        The floor of each element in `x`, with specified data type.
+
     Example::
         >>> @ti.kernel
         >>> def test():
-        >>>     x = ti.Matrix([3.14, -1.5])
-        >>>     y = ti.floor(x)
-        >>>     print(y)  # [3.0, -2.0]
+        >>>     x = ti.Matrix([-1.1, 2.2, 3.])
+        >>>     y = ti.floor(x, ti.f64)
+        >>>     print(y)  # [-2.000000000000, 2.000000000000, 3.000000000000]
     """
-    return _unary_operation(_ti_core.expr_floor, math.floor, x)
+    result = _floor(x)
+    if dtype != float:
+        result = cast(result, dtype)
```

Comment:
> Could you provide an example running into the bug? IIRC the current implementation returns a value of the same type as the input value.

@neozhaoliang @strongoier Let's simply use `None` as the default type for code clarity. `float` is the user-specified default floating-point precision and can be misleading here.

```suggestion
    if dtype != None:
        result = cast(result, dtype)
```

Reply:
Ah sorry about my misunderstanding. I was saying that there's no bug in current Taichi implementation :-)

And yes we should make `dtype=None` as default to allow users to fallback to current Taichi implementation.

Reply:
Actually there is a bug: when `float` is `ti.32`, the user calls `ti.floor(x, dtype=ti.f32)` and `x` is `ti.f64`, the user gets `ti.f64` return type instead of `ti.f32` :-)